// standardized-audio-context/src/types/add-connection-to-audio-node-factory.ts
import { TAddActiveInputConnectionToAudioNodeFunction } from './add-active-input-connection-to-audio-node-function';
import { TAddConnectionToAudioNodeFunction } from './add-connection-to-audio-node-function';
import { TAddPassiveInputConnectionToAudioNodeFunction } from './add-passive-input-connection-to-audio-node-function';
import { TConnectNativeAudioNodeToNativeAudioNodeFunction } from './connect-native-audio-node-to-native-audio-node-function';
import { TDeleteActiveInputConnectionToAudioNodeFunction } from './delete-active-input-connection-to-audio-node-function';
import { TDisconnectNativeAudioNodeFromNativeAudioNodeFunction } from './disconnect-native-audio-node-from-native-audio-node-function';
import { TGetAudioNodeConnectionsFunction } from './get-audio-node-connections-function';
import { TGetAudioNodeTailTimeFunction } from './get-audio-node-tail-time-function';
import { TGetEventListenersOfAudioNodeFunction } from './get-event-listeners-of-audio-node-function';
import { TGetNativeAudioNodeFunction } from './get-native-audio-node-function';
import { TInsertElementInSetFunction } from './insert-element-in-set-function';
import { TIsActiveAudioNodeFunction } from './is-active-audio-node-function';
import { TIsPartOfACycleFunction } from './is-part-of-a-cycle-function';
import { TIsPassiveAudioNodeFunction } from './is-passive-audio-node-function';

export type TAddConnectionToAudioNodeFactory = (
    addActiveInputConnectionToAudioNode: TAddActiveInputConnectionToAudioNodeFunction,
    addPassiveInputConnectionToAudioNode: TAddPassiveInputConnectionToAudioNodeFunction,
    connectNativeAudioNodeToNativeAudioNode: TConnectNativeAudioNodeToNativeAudioNodeFunction,
    deleteActiveInputConnectionToAudioNode: TDeleteActiveInputConnectionToAudioNodeFunction,
    disconnectNativeAudioNodeFromNativeAudioNode: TDisconnectNativeAudioNodeFromNativeAudioNodeFunction,
    getAudioNodeConnections: TGetAudioNodeConnectionsFunction,
    getAudioNodeTailTime: TGetAudioNodeTailTimeFunction,
    getEventListenersOfAudioNode: TGetEventListenersOfAudioNodeFunction,
    getNativeAudioNode: TGetNativeAudioNodeFunction,
    insertElementInSet: TInsertElementInSetFunction,
    isActiveAudioNode: TIsActiveAudioNodeFunction,
    isPartOfACycle: TIsPartOfACycleFunction,
    isPassiveAudioNode: TIsPassiveAudioNodeFunction
) => TAddConnectionToAudioNodeFunction;
