// standardized-audio-context/src/types/biquad-filter-node-constructor-factory.ts
import { TAudioNodeConstructor } from './audio-node-constructor';
import { TAudioParamFactory } from './audio-param-factory';
import { TBiquadFilterNodeConstructor } from './biquad-filter-node-constructor';
import { TBiquadFilterNodeRendererFactory } from './biquad-filter-node-renderer-factory';
import { TGetNativeContextFunction } from './get-native-context-function';
import { TInvalidAccessErrorFactory } from './invalid-access-error-factory';
import { TIsNativeOfflineAudioContextFunction } from './is-native-offline-audio-context-function';
import { TNativeBiquadFilterNodeFactory } from './native-biquad-filter-node-factory';
import { TSetAudioNodeTailTimeFunction } from './set-audio-node-tail-time-function';

export type TBiquadFilterNodeConstructorFactory = (
    audioNodeConstructor: TAudioNodeConstructor,
    createAudioParam: TAudioParamFactory,
    createBiquadFilterNodeRenderer: TBiquadFilterNodeRendererFactory,
    createInvalidAccessError: TInvalidAccessErrorFactory,
    createNativeBiquadFilterNode: TNativeBiquadFilterNodeFactory,
    getNativeContext: TGetNativeContextFunction,
    isNativeOfflineAudioContext: TIsNativeOfflineAudioContextFunction,
    setAudioNodeTailTime: TSetAudioNodeTailTimeFunction
) => TBiquadFilterNodeConstructor;
